storage/copy.go

```diff
@@ -204,45 +207,39 @@ func (c *Composer) Run(ctx context.Context) (attrs *ObjectAttrs, err error) {
 		}
 	}
 
-	// TODO: transport agnostic interface starts here.
-	req := &raw.ComposeRequest{}
-	// Compose requires a non-empty Destination, so we always set it,
-	// even if the caller-provided ObjectAttrs is the zero value.
-	req.Destination = c.ObjectAttrs.toRawObject(c.dst.bucket)
-	if c.SendCRC32C {
-		req.Destination.Crc32c = encodeUint32(c.ObjectAttrs.CRC32C)
-	}
-	for _, src := range c.srcs {
-		srcObj := &raw.ComposeRequestSourceObjects{
-			Name: src.object,
-		}
-		if err := applyConds("ComposeFrom source", src.gen, src.conds, composeSourceObj{srcObj}); err != nil {
-			return nil, err
+	req := &composeObjectRequest{
+		dstBucket: c.dst.bucket,
+		predefinedACL: c.PredefinedACL,
+		sendCRC32C: c.SendCRC32C,
+	}
+	req.dstObject = destinationObject{
+		name: c.dst.object,
+		bucket: c.dst.bucket,
+		conds: c.dst.conds,
+		attrs: &c.ObjectAttrs,
+		encryptionKey: c.dst.encryptionKey,
+	}
+	for _, src := range(c.srcs) {
+		s := sourceObject{
+			name: src.object,
+			bucket: src.bucket,
+			gen: src.gen,
+			conds: src.conds,
 		}
-		req.SourceObjects = append(req.SourceObjects, srcObj)
+		req.srcs = append(req.srcs, s)
 	}
 
-	call := c.dst.c.raw.Objects.Compose(c.dst.bucket, c.dst.object, req).Context(ctx)
-	if err := applyConds("ComposeFrom destination", c.dst.gen, c.dst.conds, call); err != nil {
-		return nil, err
+	// TODO: factor this out to a function?
```

Comment:
Is the definition of idempotency _here_ a generalized one? Or is it specific to Copy/Compose?

Reply:
It's specific to copy/compose. However was thinking I could add a helper func to create the storage opts just to eliminate a few of these lines of boilerplate which will be in every func:

```
func makeStorageOpts(idempotent bool, retry *retryConfig, userProject string) []storageOption {...}
```

Maybe this defeats the point of the options interface though?

Reply:
> Maybe this defeats the point of the options interface though?

Perhaps...let's just keep the TODO and evaluate as we go?

Reply:
I'll be adding this in a separate PR.